// @tarojs/taro-h5/dist/api/media/EditorContext.js
import { temporarilyNotSupport } from '../utils';
// 富文本
export class EditorContext {
    constructor() {
        this.blur = temporarilyNotSupport('EditorContext.blur');
        this.clear = temporarilyNotSupport('EditorContext.clear');
        this.format = temporarilyNotSupport('EditorContext.format');
        this.getContents = temporarilyNotSupport('EditorContext.getContents');
        this.insertDivider = temporarilyNotSupport('EditorContext.insertDivider');
        this.insertImage = temporarilyNotSupport('EditorContext.insertImage');
        this.insertText = temporarilyNotSupport('EditorContext.insertText');
        this.redo = temporarilyNotSupport('EditorContext.redo');
        this.removeFormat = temporarilyNotSupport('EditorContext.removeFormat');
        this.scrollIntoView = temporarilyNotSupport('EditorContext.scrollIntoView');
        this.setContents = temporarilyNotSupport('EditorContext.setContents');
        this.undo = temporarilyNotSupport('EditorContext.undo');
    }
}
