// vant/lib/grid/Grid.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  GRID_KEY: () => GRID_KEY,
  default: () => stdin_default
});
var import_vue = __toModule(require("vue"));
var import_vue2 = __toModule(require("vue"));
var import_utils = __toModule(require("../utils"));
var import_constant = __toModule(require("../utils/constant"));
var import_use = __toModule(require("@vant/use"));
const [name, bem] = (0, import_utils.createNamespace)("grid");
const gridProps = {
  square: Boolean,
  center: import_utils.truthProp,
  border: import_utils.truthProp,
  gutter: import_utils.numericProp,
  reverse: Boolean,
  iconSize: import_utils.numericProp,
  direction: String,
  clickable: Boolean,
  columnNum: (0, import_utils.makeNumericProp)(4)
};
const GRID_KEY = Symbol(name);
var stdin_default = (0, import_vue2.defineComponent)({
  name,
  props: gridProps,
  setup(props, {
    slots
  }) {
    const {
      linkChildren
    } = (0, import_use.useChildren)(GRID_KEY);
    linkChildren({
      props
    });
    return () => {
      var _a;
      return (0, import_vue.createVNode)("div", {
        "style": {
          paddingLeft: (0, import_utils.addUnit)(props.gutter)
        },
        "class": [bem(), {
          [import_constant.BORDER_TOP]: props.border && !props.gutter
        }]
      }, [(_a = slots.default) == null ? void 0 : _a.call(slots)]);
    };
  }
});
