// vant/lib/loading/Loading.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  default: () => stdin_default
});
var import_vue = __toModule(require("vue"));
var import_vue2 = __toModule(require("vue"));
var import_utils = __toModule(require("../utils"));
const [name, bem] = (0, import_utils.createNamespace)("loading");
const SpinIcon = Array(12).fill(null).map((_, index) => (0, import_vue.createVNode)("i", {
  "class": bem("line", String(index + 1))
}, null));
const CircularIcon = (0, import_vue.createVNode)("svg", {
  "class": bem("circular"),
  "viewBox": "25 25 50 50"
}, [(0, import_vue.createVNode)("circle", {
  "cx": "50",
  "cy": "50",
  "r": "20",
  "fill": "none"
}, null)]);
const loadingProps = {
  size: import_utils.numericProp,
  type: (0, import_utils.makeStringProp)("circular"),
  color: String,
  vertical: Boolean,
  textSize: import_utils.numericProp,
  textColor: String
};
var stdin_default = (0, import_vue2.defineComponent)({
  name,
  props: loadingProps,
  setup(props, {
    slots
  }) {
    const spinnerStyle = (0, import_vue2.computed)(() => (0, import_utils.extend)({
      color: props.color
    }, (0, import_utils.getSizeStyle)(props.size)));
    const renderText = () => {
      var _a;
      if (slots.default) {
        return (0, import_vue.createVNode)("span", {
          "class": bem("text"),
          "style": {
            fontSize: (0, import_utils.addUnit)(props.textSize),
            color: (_a = props.textColor) != null ? _a : props.color
          }
        }, [slots.default()]);
      }
    };
    return () => {
      const {
        type,
        vertical
      } = props;
      return (0, import_vue.createVNode)("div", {
        "class": bem([type, {
          vertical
        }])
      }, [(0, import_vue.createVNode)("span", {
        "class": bem("spinner", type),
        "style": spinnerStyle.value
      }, [type === "spinner" ? SpinIcon : CircularIcon]), renderText()]);
    };
  }
});
