// vant/lib/utils/props.js
var __defProp = Object.defineProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
__export(exports, {
  makeArrayProp: () => makeArrayProp,
  makeNumberProp: () => makeNumberProp,
  makeNumericProp: () => makeNumericProp,
  makeRequiredProp: () => makeRequiredProp,
  makeStringProp: () => makeStringProp,
  numericProp: () => numericProp,
  truthProp: () => truthProp,
  unknownProp: () => unknownProp
});
const unknownProp = null;
const numericProp = [Number, String];
const truthProp = {
  type: Boolean,
  default: true
};
const makeRequiredProp = (type) => ({
  type,
  required: true
});
const makeArrayProp = () => ({
  type: Array,
  default: () => []
});
const makeNumberProp = (defaultVal) => ({
  type: Number,
  default: defaultVal
});
const makeNumericProp = (defaultVal) => ({
  type: numericProp,
  default: defaultVal
});
const makeStringProp = (defaultVal) => ({
  type: String,
  default: defaultVal
});
