// @tarojs/plugin-platform-qq/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var pluginPlatformWeapp = require('@tarojs/plugin-platform-weapp');

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */

function __awaiter(thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

const components = {
    // ======== 调整属性 ========
    Button: {
        'app-packagename': '',
        'app-bundleid': '',
        'app-connect-id': '',
        'group-id': '',
        'public-id': '',
        'share-type': '27',
        'share-mode': '[\'qq\', \'qzone\']',
        'aria-label': '',
        'open-id': '',
        'share-message-friend-info': '',
        bindAddFriend: '',
        bindAddGroupApp: ''
    }
};

const PACKAGE_NAME = '@tarojs/plugin-platform-qq';
class QQ extends pluginPlatformWeapp.Weapp {
    /**
     * 1. setupTransaction - init
     * 2. setup
     * 3. setupTransaction - close
     * 4. buildTransaction - init
     * 5. build
     * 6. buildTransaction - close
     */
    constructor(ctx, config, pluginOptions) {
        super(ctx, config, pluginOptions);
        this.platform = 'qq';
        this.globalObject = 'qq';
        this.projectConfigJson = 'project.qq.json';
        this.runtimePath = `${PACKAGE_NAME}/dist/runtime`;
        this.fileType = {
            templ: '.qml',
            style: '.qss',
            config: '.json',
            script: '.js',
            xs: '.wxs'
        };
        this.buildTransaction.addWrapper({
            init: this.beforeBuild
        });
    }
    beforeBuild() {
        // 处理 QQ 与微信的组件差异
        this.template.mergeComponents(this.ctx, components);
        this.template.Adapter = {
            if: 'qq:if',
            else: 'qq:else',
            elseif: 'qq:elif',
            for: 'qq:for',
            forItem: 'qq:for-item',
            forIndex: 'qq:for-index',
            key: 'qq:key',
            xs: 'wxs',
            type: 'qq'
        };
    }
}

var index = (ctx, options) => {
    ctx.registerPlatform({
        name: 'qq',
        useConfigName: 'mini',
        fn({ config }) {
            return __awaiter(this, void 0, void 0, function* () {
                const program = new QQ(ctx, config, options);
                yield program.start();
            });
        }
    });
};

exports.QQ = QQ;
exports['default'] = index;
