// vant/lib/switch/Switch.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  default: () => stdin_default
});
var import_vue = __toModule(require("vue"));
var import_vue2 = __toModule(require("vue"));
var import_utils = __toModule(require("../utils"));
var import_use = __toModule(require("@vant/use"));
var import_loading = __toModule(require("../loading"));
const [name, bem] = (0, import_utils.createNamespace)("switch");
const switchProps = {
  size: import_utils.numericProp,
  loading: Boolean,
  disabled: Boolean,
  modelValue: import_utils.unknownProp,
  activeColor: String,
  inactiveColor: String,
  activeValue: {
    type: import_utils.unknownProp,
    default: true
  },
  inactiveValue: {
    type: import_utils.unknownProp,
    default: false
  }
};
var stdin_default = (0, import_vue2.defineComponent)({
  name,
  props: switchProps,
  emits: ["change", "update:modelValue"],
  setup(props, {
    emit
  }) {
    const isChecked = () => props.modelValue === props.activeValue;
    const onClick = () => {
      if (!props.disabled && !props.loading) {
        const newValue = isChecked() ? props.inactiveValue : props.activeValue;
        emit("update:modelValue", newValue);
        emit("change", newValue);
      }
    };
    const renderLoading = () => {
      if (props.loading) {
        const color = isChecked() ? props.activeColor : props.inactiveColor;
        return (0, import_vue.createVNode)(import_loading.Loading, {
          "class": bem("loading"),
          "color": color
        }, null);
      }
    };
    (0, import_use.useCustomFieldValue)(() => props.modelValue);
    return () => {
      const {
        size,
        loading,
        disabled,
        activeColor,
        inactiveColor
      } = props;
      const checked = isChecked();
      const style = {
        fontSize: (0, import_utils.addUnit)(size),
        backgroundColor: checked ? activeColor : inactiveColor
      };
      return (0, import_vue.createVNode)("div", {
        "role": "switch",
        "class": bem({
          on: checked,
          loading,
          disabled
        }),
        "style": style,
        "tabindex": disabled ? void 0 : 0,
        "aria-checked": checked,
        "onClick": onClick
      }, [(0, import_vue.createVNode)("div", {
        "class": bem("node")
      }, [renderLoading()])]);
    };
  }
});
