// vant/lib/skeleton/Skeleton.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  default: () => stdin_default
});
var import_vue = __toModule(require("vue"));
var import_vue2 = __toModule(require("vue"));
var import_utils = __toModule(require("../utils"));
const [name, bem] = (0, import_utils.createNamespace)("skeleton");
const DEFAULT_ROW_WIDTH = "100%";
const DEFAULT_LAST_ROW_WIDTH = "60%";
const skeletonProps = {
  row: (0, import_utils.makeNumericProp)(0),
  title: Boolean,
  round: Boolean,
  avatar: Boolean,
  loading: import_utils.truthProp,
  animate: import_utils.truthProp,
  avatarSize: import_utils.numericProp,
  titleWidth: import_utils.numericProp,
  avatarShape: (0, import_utils.makeStringProp)("round"),
  rowWidth: {
    type: [Number, String, Array],
    default: DEFAULT_ROW_WIDTH
  }
};
var stdin_default = (0, import_vue2.defineComponent)({
  name,
  props: skeletonProps,
  setup(props, {
    slots
  }) {
    const renderAvatar = () => {
      if (props.avatar) {
        return (0, import_vue.createVNode)("div", {
          "class": bem("avatar", props.avatarShape),
          "style": (0, import_utils.getSizeStyle)(props.avatarSize)
        }, null);
      }
    };
    const renderTitle = () => {
      if (props.title) {
        return (0, import_vue.createVNode)("h3", {
          "class": bem("title"),
          "style": {
            width: (0, import_utils.addUnit)(props.titleWidth)
          }
        }, null);
      }
    };
    const getRowWidth = (index) => {
      const {
        rowWidth
      } = props;
      if (rowWidth === DEFAULT_ROW_WIDTH && index === +props.row - 1) {
        return DEFAULT_LAST_ROW_WIDTH;
      }
      if (Array.isArray(rowWidth)) {
        return rowWidth[index];
      }
      return rowWidth;
    };
    const renderRows = () => Array(+props.row).fill("").map((_, i) => (0, import_vue.createVNode)("div", {
      "class": bem("row"),
      "style": {
        width: (0, import_utils.addUnit)(getRowWidth(i))
      }
    }, null));
    return () => {
      var _a;
      if (!props.loading) {
        return (_a = slots.default) == null ? void 0 : _a.call(slots);
      }
      return (0, import_vue.createVNode)("div", {
        "class": bem({
          animate: props.animate,
          round: props.round
        })
      }, [renderAvatar(), (0, import_vue.createVNode)("div", {
        "class": bem("content")
      }, [renderTitle(), renderRows()])]);
    };
  }
});
