// @nutui/nutui/dist/packages/_es/Form.js
/*!
* @nutui/nutui v3.1.16 Sun Jan 30 2022 10:58:10 GMT+0800 (中国标准时间)
* (c) 2021 @jdf2e.
* Released under the MIT License.
*/
import { c as createComponent } from "./component.js";
import { i as isPromise } from "./util.js";
import { computed, reactive, provide, watch, resolveComponent, openBlock, createElementBlock, createVNode, withCtx, renderSlot } from "vue";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
const component = {
  props: {
    modelValue: {
      type: Object,
      default: {}
    }
  },
  components: {},
  emits: ["validate"],
  setup(props, { emit, slots }) {
    const formErrorTip = computed(() => reactive({}));
    provide("formErrorTip", formErrorTip);
    const clearErrorTips = (value = props.modelValue) => {
      Object.keys(formErrorTip.value).forEach((item) => {
        formErrorTip.value[item] = "";
      });
    };
    const reset = () => {
      clearErrorTips();
    };
    watch(() => props.modelValue, (value) => {
      clearErrorTips(value);
    }, { immediate: true });
    const findFormItem = (vnodes) => {
      let task = [];
      vnodes.forEach((vnode, index2) => {
        var _a, _b, _c;
        let type = vnode.type;
        type = type.name || type;
        if (type == "nut-form-item") {
          task.push({
            prop: (_a = vnode.props) == null ? void 0 : _a["prop"],
            rules: ((_b = vnode.props) == null ? void 0 : _b["rules"]) || []
          });
        } else if (Array.isArray(vnode.children) && ((_c = vnode.children) == null ? void 0 : _c.length)) {
          task = task.concat(findFormItem(vnode.children));
        }
      });
      return task;
    };
    const tipMessage = (errorMsg) => {
      if (errorMsg.message) {
        emit("validate", errorMsg);
      }
      formErrorTip.value[errorMsg.prop] = errorMsg.message;
    };
    const checkRule = (item) => {
      const { rules, prop } = item;
      const _Promise = (errorMsg) => {
        return new Promise((resolve, reject) => {
          tipMessage(errorMsg);
          resolve(errorMsg);
        });
      };
      const getPropByPath = (obj, keyPath) => {
        return keyPath.split(".").reduce((prev, curr) => prev[curr], obj);
      };
      if (!prop) {
        console.warn("[NutUI] <FormItem> \u4F7F\u7528 rules \u6821\u9A8C\u89C4\u5219\u65F6 , \u5FC5\u987B\u8BBE\u7F6E prop \u53C2\u6570");
      }
      let value = getPropByPath(props.modelValue, prop || "");
      tipMessage({ prop, message: "" });
      while (rules.length) {
        const { required, validator, regex, message } = rules.shift();
        const errorMsg = { prop, message };
        if (required) {
          if (!value) {
            return _Promise(errorMsg);
          }
        }
        if (regex && !regex.test(String(value))) {
          return _Promise(errorMsg);
        }
        if (validator) {
          const result = validator(value);
          if (isPromise(result)) {
            return new Promise((r, j) => {
              result.then((res) => {
                if (!res) {
                  tipMessage(errorMsg);
                  r(errorMsg);
                } else {
                  r(true);
                }
              });
            });
          } else {
            if (!result) {
              return _Promise(errorMsg);
            }
          }
        }
      }
      return Promise.resolve(true);
    };
    const validate = (customProp = "") => {
      return new Promise((resolve, reject) => {
        let task = findFormItem(slots.default());
        let errors = task.map((item) => {
          if (customProp) {
            if (customProp == item.prop) {
              return checkRule(item);
            } else {
              return Promise.resolve(true);
            }
          } else {
            return checkRule(item);
          }
        });
        Promise.all(errors).then((errorRes) => {
          errorRes = errorRes.filter((item) => item != true);
          const res = { valid: true, errors: [] };
          if (errorRes.length) {
            res.valid = false;
            res.errors = errorRes;
          }
          resolve(res);
        });
      });
    };
    const onSubmit = () => {
      validate();
      return false;
    };
    return { validate, reset, onSubmit, formErrorTip };
  }
};
const { create } = createComponent("form");
const _sfc_main = create(component);
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_nut_cell_group = resolveComponent("nut-cell-group");
  return openBlock(), createElementBlock("form", {
    class: "nut-form",
    onSubmit: _cache[0] || (_cache[0] = (...args) => _ctx.onSubmit && _ctx.onSubmit(...args))
  }, [
    createVNode(_component_nut_cell_group, null, {
      default: withCtx(() => [
        renderSlot(_ctx.$slots, "default")
      ]),
      _: 3
    })
  ], 32);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
