// @nutui/nutui/dist/packages/_es/SubSideNavBar.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
/*!
* @nutui/nutui v3.1.16 Sun Jan 30 2022 10:58:10 GMT+0800 (中国标准时间)
* (c) 2021 @jdf2e.
* Released under the MIT License.
*/
import { reactive, computed, onMounted, toRefs, resolveComponent, openBlock, createElementBlock, normalizeClass, createElementVNode, withModifiers, toDisplayString, createVNode, normalizeStyle, renderSlot } from "vue";
import { c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
const { componentName, create } = createComponent("subsidenavbar");
const _sfc_main = create({
  props: {
    title: {
      type: String,
      default: ""
    },
    ikey: {
      type: [String, Number],
      default: ""
    },
    open: {
      type: Boolean,
      default: true
    }
  },
  emits: ["title-click"],
  setup: (props, context) => {
    const state = reactive({
      direction: ""
    });
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true
      };
    });
    const style = computed(() => {
      return {
        height: !state.direction ? "auto" : "0px"
      };
    });
    const handleClick = () => {
      context.emit("title-click");
      state.direction = !state.direction ? "up" : "";
    };
    onMounted(() => {
      state.direction = props.open ? "" : "up";
    });
    return __spreadProps(__spreadValues({}, toRefs(state)), {
      classes,
      style,
      handleClick
    });
  }
});
const _hoisted_1 = ["ikey"];
const _hoisted_2 = { class: "nut-subsidenavbar__title__text" };
const _hoisted_3 = { class: "nut-subsidenavbar__title__icon" };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_nut_icon = resolveComponent("nut-icon");
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes),
    ikey: _ctx.ikey
  }, [
    createElementVNode("view", {
      class: "nut-subsidenavbar__title",
      onClick: _cache[0] || (_cache[0] = withModifiers((...args) => _ctx.handleClick && _ctx.handleClick(...args), ["stop"]))
    }, [
      createElementVNode("span", _hoisted_2, toDisplayString(_ctx.title), 1),
      createElementVNode("span", _hoisted_3, [
        createVNode(_component_nut_icon, {
          name: "down-arrow",
          class: normalizeClass(_ctx.direction)
        }, null, 8, ["class"])
      ])
    ]),
    createElementVNode("view", {
      class: normalizeClass(["nut-subsidenavbar__list", !_ctx.direction ? "nutFadeIn" : "nutFadeOut"]),
      style: normalizeStyle(_ctx.style)
    }, [
      renderSlot(_ctx.$slots, "default")
    ], 6)
  ], 10, _hoisted_1);
}
var index = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index as default };
