// vant/lib/nav-bar/NavBar.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  default: () => stdin_default
});
var import_vue = __toModule(require("vue"));
var import_vue2 = __toModule(require("vue"));
var import_utils = __toModule(require("../utils"));
var import_use_placeholder = __toModule(require("../composables/use-placeholder"));
var import_icon = __toModule(require("../icon"));
const [name, bem] = (0, import_utils.createNamespace)("nav-bar");
const navBarProps = {
  title: String,
  fixed: Boolean,
  zIndex: import_utils.numericProp,
  border: import_utils.truthProp,
  leftText: String,
  rightText: String,
  leftArrow: Boolean,
  placeholder: Boolean,
  safeAreaInsetTop: Boolean
};
var stdin_default = (0, import_vue2.defineComponent)({
  name,
  props: navBarProps,
  emits: ["click-left", "click-right"],
  setup(props, {
    emit,
    slots
  }) {
    const navBarRef = (0, import_vue2.ref)();
    const renderPlaceholder = (0, import_use_placeholder.usePlaceholder)(navBarRef, bem);
    const onClickLeft = (event) => emit("click-left", event);
    const onClickRight = (event) => emit("click-right", event);
    const renderLeft = () => {
      if (slots.left) {
        return slots.left();
      }
      return [props.leftArrow && (0, import_vue.createVNode)(import_icon.Icon, {
        "class": bem("arrow"),
        "name": "arrow-left"
      }, null), props.leftText && (0, import_vue.createVNode)("span", {
        "class": bem("text")
      }, [props.leftText])];
    };
    const renderRight = () => {
      if (slots.right) {
        return slots.right();
      }
      return (0, import_vue.createVNode)("span", {
        "class": bem("text")
      }, [props.rightText]);
    };
    const renderNavBar = () => {
      const {
        title,
        fixed,
        border,
        zIndex
      } = props;
      const style = (0, import_utils.getZIndexStyle)(zIndex);
      const hasLeft = props.leftArrow || props.leftText || slots.left;
      const hasRight = props.rightText || slots.right;
      return (0, import_vue.createVNode)("div", {
        "ref": navBarRef,
        "style": style,
        "class": [bem({
          fixed,
          "safe-area-inset-top": props.safeAreaInsetTop
        }), {
          [import_utils.BORDER_BOTTOM]: border
        }]
      }, [(0, import_vue.createVNode)("div", {
        "class": bem("content")
      }, [hasLeft && (0, import_vue.createVNode)("div", {
        "class": [bem("left"), import_utils.HAPTICS_FEEDBACK],
        "onClick": onClickLeft
      }, [renderLeft()]), (0, import_vue.createVNode)("div", {
        "class": [bem("title"), "van-ellipsis"]
      }, [slots.title ? slots.title() : title]), hasRight && (0, import_vue.createVNode)("div", {
        "class": [bem("right"), import_utils.HAPTICS_FEEDBACK],
        "onClick": onClickRight
      }, [renderRight()])])]);
    };
    return () => {
      if (props.fixed && props.placeholder) {
        return renderPlaceholder(renderNavBar);
      }
      return renderNavBar();
    };
  }
});
