// vant/lib/uploader/UploaderPreviewItem.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, copyDefault, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && (copyDefault || key !== "default"))
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toCommonJS = /* @__PURE__ */ ((cache) => {
  return (module2, temp) => {
    return cache && cache.get(module2) || (temp = __reExport(__markAsModule({}), module2, 1), cache && cache.set(module2, temp), temp);
  };
})(typeof WeakMap !== "undefined" ? /* @__PURE__ */ new WeakMap() : 0);
var stdin_exports = {};
__export(stdin_exports, {
  default: () => stdin_default
});
var import_vue = require("vue");
var import_vue2 = require("vue");
var import_utils = require("./utils");
var import_utils2 = require("../utils");
var import_icon = require("../icon");
var import_image = require("../image");
var import_loading = require("../loading");
var stdin_default = (0, import_vue2.defineComponent)({
  props: {
    name: import_utils2.numericProp,
    item: (0, import_utils2.makeRequiredProp)(Object),
    index: Number,
    imageFit: String,
    lazyLoad: Boolean,
    deletable: Boolean,
    previewSize: import_utils2.numericProp,
    beforeDelete: Function
  },
  emits: ["delete", "preview"],
  setup(props, {
    emit,
    slots
  }) {
    const renderMask = () => {
      const {
        status,
        message
      } = props.item;
      if (status === "uploading" || status === "failed") {
        const MaskIcon = status === "failed" ? (0, import_vue.createVNode)(import_icon.Icon, {
          "name": "close",
          "class": (0, import_utils.bem)("mask-icon")
        }, null) : (0, import_vue.createVNode)(import_loading.Loading, {
          "class": (0, import_utils.bem)("loading")
        }, null);
        const showMessage = (0, import_utils2.isDef)(message) && message !== "";
        return (0, import_vue.createVNode)("div", {
          "class": (0, import_utils.bem)("mask")
        }, [MaskIcon, showMessage && (0, import_vue.createVNode)("div", {
          "class": (0, import_utils.bem)("mask-message")
        }, [message])]);
      }
    };
    const onDelete = (event) => {
      const {
        name,
        item,
        index,
        beforeDelete
      } = props;
      event.stopPropagation();
      (0, import_utils2.callInterceptor)(beforeDelete, {
        args: [item, {
          name,
          index
        }],
        done: () => emit("delete")
      });
    };
    const onPreview = () => emit("preview");
    const renderDeleteIcon = () => {
      if (props.deletable && props.item.status !== "uploading") {
        return (0, import_vue.createVNode)("div", {
          "role": "button",
          "class": (0, import_utils.bem)("preview-delete"),
          "tabindex": 0,
          "aria-label": (0, import_utils.t)("delete"),
          "onClick": onDelete
        }, [(0, import_vue.createVNode)(import_icon.Icon, {
          "name": "cross",
          "class": (0, import_utils.bem)("preview-delete-icon")
        }, null)]);
      }
    };
    const renderCover = () => {
      if (slots["preview-cover"]) {
        const {
          index,
          item
        } = props;
        return (0, import_vue.createVNode)("div", {
          "class": (0, import_utils.bem)("preview-cover")
        }, [slots["preview-cover"]((0, import_utils2.extend)({
          index
        }, item))]);
      }
    };
    const renderPreview = () => {
      const {
        item
      } = props;
      if ((0, import_utils.isImageFile)(item)) {
        return (0, import_vue.createVNode)(import_image.Image, {
          "fit": props.imageFit,
          "src": item.content || item.url,
          "class": (0, import_utils.bem)("preview-image"),
          "width": props.previewSize,
          "height": props.previewSize,
          "lazyLoad": props.lazyLoad,
          "onClick": onPreview
        }, {
          default: renderCover
        });
      }
      return (0, import_vue.createVNode)("div", {
        "class": (0, import_utils.bem)("file"),
        "style": (0, import_utils2.getSizeStyle)(props.previewSize)
      }, [(0, import_vue.createVNode)(import_icon.Icon, {
        "class": (0, import_utils.bem)("file-icon"),
        "name": "description"
      }, null), (0, import_vue.createVNode)("div", {
        "class": [(0, import_utils.bem)("file-name"), "van-ellipsis"]
      }, [item.file ? item.file.name : item.url]), renderCover()]);
    };
    return () => (0, import_vue.createVNode)("div", {
      "class": (0, import_utils.bem)("preview")
    }, [renderPreview(), renderMask(), renderDeleteIcon()]);
  }
});
module.exports = __toCommonJS(stdin_exports);
